File: 74-search-a-2d-matrix/74-search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        m = len(matrix)
        n = len(matrix[0])
        
        low, high = 0, m
        
        # Finding the row of target
        while low < high:
            mid = low + (high - low) // 2
            
            if matrix[mid][0] > target:
                high = mid
            else:
                low = mid + 1
        
        row = low - 1
        low, high = 0, n
        
        # Finding column of target
        while low < high:
            mid = low + (high - low) // 2
            
            if matrix[row][mid] >= target:
                high = mid
            else:
                low = mid + 1
            
        col = low
        
        return (col < n) and (matrix[row][col] == target)

File: 74-search-a-2d-matrix/74-search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m = matrix.size();
        int n = matrix[0].size();
        
        int low = 0 , high = m * n;
        
        while(low < high){
            int mid = low + (high - low) / 2;
            int i = mid / n, j = mid % n;
            
            if(matrix[i][j] >= target) high = mid;
            else low = mid + 1;
        }
        return (low < m * n && matrix[low / n][low % n] == target);        

    }
};